mirrord/intproxy/src/proxies/incoming/interceptor.rs

```diff
@@ -0,0 +1,557 @@
+#[derive(Error, Debug)]
+pub enum Error {
+    /// IO failed.
+    #[error("io failed: {0}")]
+    Io(#[from] io::Error),
+    /// Hyper failed.
+    #[error("hyper failed: {0}")]
+    Hyper(#[from] hyper::Error),
+    /// The layer closed connection too soon to send a request.
+    #[error("connection closed too soon")]
+    ConnectionClosedTooSoon(HttpRequestFallback),
+    /// Received a request with an unsupported HTTP version.
+    #[error("{0:?} is not supported")]
+    UnsupportedHttpVersion(Version),
+    /// Occurs when the [`Interceptor`] is acting as an HTTP gateway and receives
+    /// [`MessageIn::Raw`], but the HTTP version used does not support upgrading connections.
+    #[error("upgrading connections not supported in {0:?}")]
+    UpgradeNotSupported(Version),
+    /// Occurs when reclaiming [`TcpStream`] from the HTTP connection is not possible, because a
+    /// background [`tokio::task`] panicked.
+    #[error("task for keeping http connection alive panicked")]
+    HttpConnectionTaskPanicked,
+}
+
+pub type Result<T, E = Error> = core::result::Result<T, E>;
+
+/// Manages a single intercepted connection.
+/// Multiple instances are run as [`BackgroundTask`]s by one [`IncomingProxy`](super::IncomingProxy)
+/// to manage individual connections.
+///
+/// This interceptor can proxy both raw TCP data and HTTP messages in the same TCP connection.
+/// When it receives [`MessageIn::Raw`], it starts acting as a simple proxy.
+/// When it received [`MessageIn::Http`], it starts acting as an HTTP gateway.
+pub struct Interceptor {
+    socket: TcpSocket,
+    peer: SocketAddr,
+}
+
+impl Interceptor {
+    /// Creates a new instance. When run, this instance will use the given `socket` (must be already
+    /// bound) to communicate with the given `peer`.
+    ///
+    /// # Note
+    ///
+    /// The socket can be replaced when retrying HTTP requests.
+    pub fn new(socket: TcpSocket, peer: SocketAddr) -> Self {
+        Self { socket, peer }
+    }
+}
+
+impl BackgroundTask for Interceptor {
+    type Error = Error;
+    type MessageIn = MessageIn;
+    type MessageOut = MessageOut;
+
+    async fn run(self, message_bus: &mut MessageBus<Self>) -> Result<(), Self::Error> {
+        /// Just to allow having one variable in the loop.
+        enum GenericConnection {
+            Raw(RawConnection),
+            Http(HttpConnection),
+        }
+
+        let stream = self.socket.connect(self.peer).await?;
+        let mut conn = GenericConnection::Raw(RawConnection { stream });
+
+        loop {
+            let new_conn = match conn {
+                GenericConnection::Raw(raw_conn) => raw_conn
+                    .run(message_bus)
+                    .await?
+                    .map(GenericConnection::Http),
+                GenericConnection::Http(http_conn) => http_conn
+                    .run(message_bus)
+                    .await?
+                    .map(GenericConnection::Raw),
+            };
```

Comment:
Wait, I don't get this.

You create a binding `conn = GenericConnection::Raw(RawConnection { stream })`, then `match conn`? Why are you `match`ing if it's guaranteed to be `RawConnection` (also, why even have the `GenericConnection` enum at all?

Reply:
It's in a loop :eyes:

Reply:
Sometimes this `conn` variable is `RawConnection`, sometimes it's `HttpConnection`.

Reply:
Oh, that's why it's `mut`.

I always struggle with patterns like this :face_with_spiral_eyes:

Reply:
Yeah, I tried to implement `async fn run(&mut self, ...)` on `GenericConnection`, but I need to move field values when converting. Couldn't work it out :<